### ISSUE TITLE
Add PostgreSQL 16/17 support

### ISSUE THREAD

**Issue by Vonng** [tvondra/quantile#15]
Got compile error against PostgreSQL 16/17

```
$ rpmbuild --with debuginfo --define "pgmajorversion 16" -ba ~/rpmbuild/SPECS/quantile.spec
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.Yq1UTX
+ umask 022
+ cd /home/vagrant/rpmbuild/BUILD
+ cd /home/vagrant/rpmbuild/BUILD
+ rm -rf quantile-1.1.7
+ /usr/bin/gzip -dc /home/vagrant/rpmbuild/SOURCES/quantile-1.1.7.tar.gz
+ /usr/bin/tar -xof -

+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd quantile-1.1.7
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.sgQCAL
+ umask 022
+ cd /home/vagrant/rpmbuild/BUILD
+ cd quantile-1.1.7
+ PATH=/usr/pgsql-16/bin:/home/vagrant/.cargo/bin:/home/vagrant/.local/bin:/home/vagrant/bin:/usr/share/Modules/bin:/usr/lib64/ccache:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
+ /usr/bin/make -j10
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wshadow=compatible-local -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -fPIC -fvisibility=hidden -fPIC -fvisibility=hidden -I. -I./ -I/usr/pgsql-16/include/server -I/usr/pgsql-16/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include  -c -o quantile.o quantile.c
/usr/bin/clang -Wno-ignored-attributes -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-unused-command-line-argument -Wno-compound-token-split-by-macro -O2  -I. -I./ -I/usr/pgsql-16/include/server -I/usr/pgsql-16/include/internal  -D_GNU_SOURCE -I/usr/include/libxml2  -I/usr/include -flto=thin -emit-llvm -c -o quantile.bc quantile.c
quantile.c:1058:24: error: incompatible integer to pointer conversion passing 'Datum' (aka 'unsigned long') to parameter of type 'const void *' [-Wint-conversion]
 1058 |         PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
      |                               ^~~~~~~~~~~~~~~~~~~~~~~
 1059 |                                                                                   CurrentMemoryContext));
      |                                                                                   ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/utils/array.h:265:54: note: expanded from macro 'PG_RETURN_ARRAYTYPE_P'
  265 | #define PG_RETURN_ARRAYTYPE_P(x)          PG_RETURN_POINTER(x)
      |                                                             ^
/usr/pgsql-16/include/server/fmgr.h:361:53: note: expanded from macro 'PG_RETURN_POINTER'
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
/usr/pgsql-16/include/server/postgres.h:322:29: note: passing argument to parameter 'X' here
  322 | PointerGetDatum(const void *X)
      |                             ^
quantile.c:1078:24: error: incompatible integer to pointer conversion passing 'Datum' (aka 'unsigned long') to parameter of type 'const void *' [-Wint-conversion]
 1078 |         PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
      |                               ^~~~~~~~~~~~~~~~~~~~~~~
 1079 |                                                                                   CurrentMemoryContext));
      |                                                                                   ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/utils/array.h:265:54: note: expanded from macro 'PG_RETURN_ARRAYTYPE_P'
  265 | #define PG_RETURN_ARRAYTYPE_P(x)          PG_RETURN_POINTER(x)
      |                                                             ^
/usr/pgsql-16/include/server/fmgr.h:361:53: note: expanded from macro 'PG_RETURN_POINTER'
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
/usr/pgsql-16/include/server/postgres.h:322:29: note: passing argument to parameter 'X' here
  322 | PointerGetDatum(const void *X)
      |                             ^
quantile.c:1099:24: error: incompatible integer to pointer conversion passing 'Datum' (aka 'unsigned long') to parameter of type 'const void *' [-Wint-conversion]
 1099 |         PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
      |                               ^~~~~~~~~~~~~~~~~~~~~~~
 1100 |                                                                                   CurrentMemoryContext));
      |                                                                                   ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/utils/array.h:265:54: note: expanded from macro 'PG_RETURN_ARRAYTYPE_P'
  265 | #define PG_RETURN_ARRAYTYPE_P(x)          PG_RETURN_POINTER(x)
      |                                                             ^
/usr/pgsql-16/include/server/fmgr.h:361:53: note: expanded from macro 'PG_RETURN_POINTER'
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
/usr/pgsql-16/include/server/postgres.h:322:29: note: passing argument to parameter 'X' here
  322 | PointerGetDatum(const void *X)
      |                             ^
quantile.c:1119:24: error: incompatible integer to pointer conversion passing 'Datum' (aka 'unsigned long') to parameter of type 'const void *' [-Wint-conversion]
 1119 |         PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
      |                               ^~~~~~~~~~~~~~~~~~~~~~~
 1120 |                                                                                   CurrentMemoryContext));
      |                                                                                   ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/utils/array.h:265:54: note: expanded from macro 'PG_RETURN_ARRAYTYPE_P'
  265 | #define PG_RETURN_ARRAYTYPE_P(x)          PG_RETURN_POINTER(x)
      |                                                             ^
/usr/pgsql-16/include/server/fmgr.h:361:53: note: expanded from macro 'PG_RETURN_POINTER'
  361 | #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
      |                                                     ^
/usr/pgsql-16/include/server/postgres.h:322:29: note: passing argument to parameter 'X' here
  322 | PointerGetDatum(const void *X)
      |                             ^
4 errors generated.
make: *** [/usr/pgsql-16/lib/pgxs/src/makefiles/../../src/Makefile.global:1093: quantile.bc] Error 1
make: *** Waiting for unfinished jobs....
In file included from /usr/pgsql-16/include/server/utils/array.h:64,
                 from quantile.c:15:
quantile.c: In function 'double_to_array':
quantile.c:1058:24: warning: passing argument 1 of 'PointerGetDatum' makes pointer from integer without a cast [-Wint-conversion]
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
                        ^~~~~~~~~~~~~~~~~~~~~~~
             CurrentMemoryContext));
             ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/fmgr.h:361:53: note: in definition of macro 'PG_RETURN_POINTER'
 #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
                                                     ^
quantile.c:1058:2: note: in expansion of macro 'PG_RETURN_ARRAYTYPE_P'
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
  ^~~~~~~~~~~~~~~~~~~~~
In file included from quantile.c:14:
/usr/pgsql-16/include/server/postgres.h:322:29: note: expected 'const void *' but argument is of type 'Datum' {aka 'long unsigned int'}
 PointerGetDatum(const void *X)
                 ~~~~~~~~~~~~^
In file included from /usr/pgsql-16/include/server/utils/array.h:64,
                 from quantile.c:15:
quantile.c: In function 'int32_to_array':
quantile.c:1078:24: warning: passing argument 1 of 'PointerGetDatum' makes pointer from integer without a cast [-Wint-conversion]
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
                        ^~~~~~~~~~~~~~~~~~~~~~~
             CurrentMemoryContext));
             ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/fmgr.h:361:53: note: in definition of macro 'PG_RETURN_POINTER'
 #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
                                                     ^
quantile.c:1078:2: note: in expansion of macro 'PG_RETURN_ARRAYTYPE_P'
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
  ^~~~~~~~~~~~~~~~~~~~~
In file included from quantile.c:14:
/usr/pgsql-16/include/server/postgres.h:322:29: note: expected 'const void *' but argument is of type 'Datum' {aka 'long unsigned int'}
 PointerGetDatum(const void *X)
                 ~~~~~~~~~~~~^
In file included from /usr/pgsql-16/include/server/utils/array.h:64,
                 from quantile.c:15:
quantile.c: In function 'int64_to_array':
quantile.c:1099:24: warning: passing argument 1 of 'PointerGetDatum' makes pointer from integer without a cast [-Wint-conversion]
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
                        ^~~~~~~~~~~~~~~~~~~~~~~
             CurrentMemoryContext));
             ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/fmgr.h:361:53: note: in definition of macro 'PG_RETURN_POINTER'
 #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
                                                     ^
quantile.c:1099:2: note: in expansion of macro 'PG_RETURN_ARRAYTYPE_P'
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
  ^~~~~~~~~~~~~~~~~~~~~
In file included from quantile.c:14:
/usr/pgsql-16/include/server/postgres.h:322:29: note: expected 'const void *' but argument is of type 'Datum' {aka 'long unsigned int'}
 PointerGetDatum(const void *X)
                 ~~~~~~~~~~~~^
In file included from /usr/pgsql-16/include/server/utils/array.h:64,
                 from quantile.c:15:
quantile.c: In function 'numeric_to_array':
quantile.c:1119:24: warning: passing argument 1 of 'PointerGetDatum' makes pointer from integer without a cast [-Wint-conversion]
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
                        ^~~~~~~~~~~~~~~~~~~~~~~
             CurrentMemoryContext));
             ~~~~~~~~~~~~~~~~~~~~~
/usr/pgsql-16/include/server/fmgr.h:361:53: note: in definition of macro 'PG_RETURN_POINTER'
 #define PG_RETURN_POINTER(x) return PointerGetDatum(x)
                                                     ^
quantile.c:1119:2: note: in expansion of macro 'PG_RETURN_ARRAYTYPE_P'
  PG_RETURN_ARRAYTYPE_P(makeArrayResult(astate,
  ^~~~~~~~~~~~~~~~~~~~~
In file included from quantile.c:14:
/usr/pgsql-16/include/server/postgres.h:322:29: note: expected 'const void *' but argument is of type 'Datum' {aka 'long unsigned int'}
 PointerGetDatum(const void *X)
                 ~~~~~~~~~~~~^
error: Bad exit status from /var/tmp/rpm-tmp.sgQCAL (%build)


RPM build errors:
    Bad exit status from /var/tmp/rpm-tmp.sgQCAL (%build)
```

**Comment by tvondra** [comment]
Should be fixed by 263db240d1ec58177bd0cd5faf084d9091582472, can you try that? I haven't bumped the version yet, but there's no functional change so just installing 1.1.7 from the git HEAD should work.

**Comment by Vonng** [comment]
Compiling from the HEAD has resolved the issue. Thanks for your response. 😄
